### PR TITLE
Fix hard-to-find search error

### LINKED ISSUE #7650: Not unique table/alias: 'm'
#### Description

Not unique table/alias: 'm'
File: C:\wamp64\www\SMF2.1\Sources\Search.php
Line: 1462

INSERT IGNORE INTO smf_tmp_log_search_topics
(id_topic)
SELECT t.id_topic
FROM smf_topics AS t
INNER JOIN smf_log_search_subjects AS subj1 ON (subj1.id_topic = t.id_topic)
INNER JOIN smf_log_search_subjects AS subj2 ON (subj2.id_topic = subj1.id_topic)
INNER JOIN smf_messages AS m ON (m.id_msg = t.id_first_msg)
INNER JOIN smf_messages AS m ON (m.id_msg = t.id_first_msg)
LEFT JOIN smf_log_search_subjects AS subj3 ON (subj3.word LIKE '%sit%' AND subj3.id_topic = t.id_topic)
LEFT JOIN smf_log_search_subjects AS subj4 ON (subj4.word LIKE '%dolor%' AND subj4.id_topic = t.id_topic)
WHERE t.id_redirect_topic = 0
AND t.redirect_expires = 0
AND subj1.word LIKE '%ipsum%'
AND subj2.word LIKE '%amet%'
AND (subj3.word IS NULL)
AND m.body NOT LIKE '%sit%'
AND (subj4.word IS NULL)
AND m.body NOT LIKE '%dolor%'
AND (m.id_member IN (1) OR (m.id_member = 0 AND (m.poster_name LIKE 'live627')))
AND t.id_board != 54
LIMIT 1200

### Steps to reproduce
1. Create a search with both an excluded word AND a username
2. No special indexes required

### Additional information/references
Bonus: Excluded word is highlighted https://www.simplemachines.org/community/index.php?action=search2;params=eJwtUVtqxDAMvErpT3-mYMnv0yx5GLYlTRYnu0shh--oLMRYGs2MJGeYH8M6tfn8OOV8P-976_utTUz36_Bcx2To2K2uHqIRxfEKCBkaEzRUlAJ1GYUIOVWRI8STVqgIhTfzWqEJkgPEVVQikiFJoeKpC_CCwCIdGTte0Eq5wJuMXGrUKWO6VlYjPdgvOXgPI_iCWJErPH1yQSZZrBHH1Ay2VhVOQ3O1UxFYLWbJYYUnkelYisQcxwoKD9qqcgufwS8ikEgpxAbgEwibSmT_SBJXT6CAeXSIbGdS26RyNUe1LaWoHsFotkfhE-_X7XmZtp_b0o7GxzboPn636bhs6_L7QrZ-MOptaf9_7QVd5q9OeG77ZEgb-nRl3nrf-tvnuB1_55556g..

### LINKED ISSUE #7654: Word boundaries have changed in MySQL 8.0.4, affecting some searches
#### Description
MySQL 8.0.4 changed the implementation of its regex engine from Henry Spencer's implementation to International Components for Unicode (ICU). This involves some non-backward compatible changes, which are listed [in the documentation](https://dev.mysql.com/doc/refman/8.0/en/regexp.html#regexp-compatibility).

> The Spencer library supports word-beginning and word-end boundary markers (`[[:<:]]` and `[[:>:]]` notation). ICU does not. For ICU, you can use `\b` to match word boundaries; double the backslash because MySQL interprets it as the escape character within strings.

@albertlast is there a function to detect MySQL version? Could you create a pull request that adds a new function to Search.php that returns correct word boundaries for each database version? I can then add a commit to it that applies that new function to the relevant code, if you're okay with that.

### Steps to reproduce
1. Enable the option to search whole words
2. Create a search for whatever
3. > Hello darkness, my old friend
   > I've come to talk with you again
   > Because a vision softly creeping
   > Left its seeds while I was sleeping
   > And the vision that was planted in my brain
   > Still remains
   > Within the sound of silence

### Environment (complete as necessary)
- Version/Git revision: I like burritos
- Database Type: MySQL
- Database Version: 8
- PHP Version: Not relevant

### Additional information/references
https://www.simplemachines.org/community/index.php?topic=584449.0
https://stackoverflow.com/questions/59998409/error-code-3685-illegal-argument-to-a-regular-expression

### PR DESCRIPTION
**Pull request by live627**
Fix #7650

### CODE REVIEW

File: Sources/Search.php [hunk 1]
```diff
@@ -688,14 +688,34 @@ function PlushSearch2()
...
 
 	$no_regexp = preg_match('~&#(?:\d{1,7}|x[0-9a-fA-F]{1,6});~', $stripped_query) === 1;
+	$is_search_regex = !empty($modSettings['search_match_words']) || !$no_regexp;
+
+	// Specify the function to search with. Regex is for word boundaries.
+	$regex_keyword = $smcFunc['db_title'] === POSTGRE_TITLE ? '~*' : 'REGEXP';
+	$query_match_type = $is_search_regex ? $regex_keyword : 'LIKE';
+	db_extend();
+	$version = $smcFunc['db_get_version']();
+	$supports_pcre = $smcFunc['db_title'] === POSTGRE_TITLE || version_compare($version, strpos($version, 'MariaDB') !== false ? '10.0.5' : '8.0.4', '>=');
```
Comment (albertlast) [inline Sources/Search.php]: This logic should be in the db layer:
https://github.com/SimpleMachines/SMF/blob/release-2.1/Sources/DbSearch-mysql.php
https://github.com/SimpleMachines/SMF/blob/release-2.1/Sources/DbSearch-postgresql.php

File: Sources/DbSearch-postgresql.php [hunk 2]
```diff
@@ -77,10 +78,22 @@ function smf_db_search_query($identifier, $db_string, $db_values = array(), $con
...
 		),
 		'insert_log_search_results_subject' => array(
-			'~LIKE~i' => 'iLIKE',
-			'~NOT\sLIKE~i' => 'NOT iLIKE',
-			'~NOT\sRLIKE~i' => '!~*',
-			'~RLIKE~i' => '~*',
+			'/\bLIKE\b/i' => 'ILIKE',
+			'/NOT\sLIKE/i' => 'NOT ILIKE',
+			'/RLIKE/i' => '/*',
+			'/NOT RLIKE/' => '!/',
```
Comment (albertlast) [inline Sources/DbSearch-postgresql.php]: /* and I/ doesn't look like valid operator: https://www.postgresql.org/docs/current/functions-matching.html

File: Sources/Search.php (hunk 1 above)
Comment (live627) [reply Sources/Search.php]: and done

File: Sources/DbSearch-postgresql.php (hunk 2 above)
Comment (live627) [reply Sources/DbSearch-postgresql.php]: oh right, I'll need to fix that

**Comment by live627** [comment]
Similar code with word boundaries is also in both `SearchAPI-Custom.php` and in `SearchAPI-Fulltext.php`. I'll need to address those.

**Comment by live627** [comment]
Is now ready for review. Searching seems to work as expected, even with excluded words.

**Pull request merged**